promql/engine.go

```diff
@@ -1325,16 +1395,100 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, storage.Warnings) {
 			newEv.interval = ev.noStepSubqueryIntervalFn(rangeMillis)
 		}
 
-		// Start with the first timestamp after (ev.startTimestamp - offset - range)
+		subqueryEvalTime := timestamp.FromFloatSeconds(e.Timestamp)
+		subqueryStartTime := ev.startTimestamp - offsetMillis - rangeMillis
+		if subqueryEvalTime != 0 {
+			newEv.endTimestamp = subqueryEvalTime - offsetMillis
+			subqueryStartTime = newEv.endTimestamp - rangeMillis
+		}
+
+		// Start with the first timestamp after subqueryStartTime
 		// that is aligned with the step (multiple of 'newEv.interval').
-		newEv.startTimestamp = newEv.interval * ((ev.startTimestamp - offsetMillis - rangeMillis) / newEv.interval)
-		if newEv.startTimestamp < (ev.startTimestamp - offsetMillis - rangeMillis) {
+		newEv.startTimestamp = newEv.interval * (subqueryStartTime / newEv.interval)
+		if newEv.startTimestamp < subqueryStartTime {
 			newEv.startTimestamp += newEv.interval
 		}
 
 		res, ws := newEv.eval(e.Expr)
 		ev.currentSamples = newEv.currentSamples
 		return res, ws
+	case *parser.StepInvariantExpr:
+		switch ce := e.Expr.(type) {
+		case *parser.StringLiteral, *parser.NumberLiteral:
+			return ev.eval(ce)
+		}
+
+		if e.Result != nil {
+			return e.Result, nil
+		}
+		newEv := &evaluator{
+			startTimestamp:           ev.startTimestamp,
+			endTimestamp:             ev.startTimestamp,
+			interval:                 ev.interval,
+			ctx:                      ev.ctx,
+			currentSamples:           ev.currentSamples,
+			maxSamples:               ev.maxSamples,
+			logger:                   ev.logger,
+			lookbackDelta:            ev.lookbackDelta,
+			noStepSubqueryIntervalFn: ev.noStepSubqueryIntervalFn,
+		}
+
+		canExtrapolate := canExtrapolateStepInvariantExpr(e.Expr)
```

Comment:
This canExtrapolate stuff still needs to be removed.

If the following logic can't be applied then something is very, very wrong.

Reply:
For matrix selectors and subqueries, we cannot apply this extrapolation stuff. Or is it that we do not expect a step invariant node for them for a range query at all?

Reply:
It is simpler to just not wrap vector and matrix selector with the step invariant node since that doesnt buy us much. But for subqueries, we still need this avoiding of extrapolation

Reply:
There is no extrapolation going on here, I don't understand why you keep on using the word "extrapolation".

One way or the other, this code needs to not exist.

Reply:
> There is no extrapolation going on here, I don't understand why you keep on using the word "extrapolation".

I can change the wording, it is "duplicating" results.

> One way or the other, this code needs to not exist.

Do you mean execute the step invariant expression for _every step_?

Reply:
> Do you mean execute the step invariant expression for every step?

No, we should only need to evaluate it once. But I don't see the need for this predicate.

Reply:
I will re-check about the predicate